### ISSUE TITLE
Better accomodate installing APK on physical device when existing app cannot be replaced

### ISSUE THREAD

**Issue by rmartin16** [beeware/briefcase#1362]
### Describe the bug

The current APK install process can error if an existing version of the app cannot be replaced. For me, this happened because the existing app on the device was installed from a different computer.

```
[helloworld] Starting app on SM G960U (5759363737573398) (device ID 5759363737573398)

[helloworld] Installing app...

>>> Running Command:
>>>     /home/russell/.cache/briefcase/tools/android_sdk/platform-tools/adb -s 5759363737573398 shell am force-stop com.example.helloworld
>>> Working Directory:
>>>     /home/russell/tmp/beeware/helloworld
>>> Return code: 0
Stopping old versions of the app... done

>>> Running Command:
>>>     /home/russell/.cache/briefcase/tools/android_sdk/platform-tools/adb -s 5759363737573398 install -r /home/russell/tmp/beeware/helloworld/build/helloworld/android/gradle/app/build/outputs/apk/debug/app-debug.apk
>>> Working Directory:
>>>     /home/russell/tmp/beeware/helloworld
>>> Command Output:
>>>     adb: failed to install /home/russell/tmp/beeware/helloworld/build/helloworld/android/gradle/app/build/outputs/apk/debug/app-debug.apk: Failure [INSTALL_FAILED_UPDATE_INCOMPATIBLE: Package com.example.helloworld signatures do not match previously installed version; ignoring!]
>>>     Performing Streamed Install
>>> Return code: 1
Installing new app version...

Unable to install APK /home/russell/tmp/beeware/helloworld/build/helloworld/android/gradle/app/build/outputs/apk/debug/app-debug.apk on 5759363737573398
```

### Steps to reproduce

1. Install an APK on a physical device
2. Attempt to install APK from a different machine or using a different identity

### Expected behavior

Fallback to an uninstall and install workflow or at least provide better information to the user about what's wrong.

A simple remedy is `~/.cache/briefcase/tools/android_sdk/platform-tools/adb uninstall "com.example.helloworld"

### Environment

- Operating System: debian 11
- Python version: 3.10.6
- Software versions:
  - Briefcase: `0.3.16.dev9+gb6057a0d`

### Logs

[briefcase.2023_07_14-17_02_12.run.log](https://github.com/beeware/briefcase/files/12053969/briefcase.2023_07_14-17_02_12.run.log)

**Comment by mhsmith** [comment]
> Fallback to an uninstall and install workflow or at least provide better information to the user about what's wrong.

I think the message is already clear enough, but falling back to a separate uninstall and install would still be useful. However, I think we should ask for confirmation first, as Android Studio does, with a warning that uninstalling the existing app will wipe out all of its data.

A similar situation could happen with the `INSTALL_FAILED_VERSION_DOWNGRADE` error.

**Comment by mhsmith** [comment]
> I think the message is already clear enough

Actually, as we've just found out from a user on Discord, the message is not displayed in non-verbose mode. All Briefcase displays in the console is:
```
Installing new app version...

Unable to install APK
```

If we'd displayed the output of the failing command, at least the user would be able to find a solution by searching for its text. As it is, there isn't much they can do except come and ask us for help, which wastes both their time and ours. 

Yes, they can find the output in the log file, but it's buried in the middle of the file between all the successful steps and the extra-verbose stack trace, so its relevance is not obvious from a user's point of view.

I know we discussed the output of failing commands last year, but I don't know the current situation. In general, I think that if a Briefcase command fails because of a subprocess, its output should always be shown to the user unless we can be certain that Briefcase has a more useful message to display instead.

**Comment by freakboy3742** [comment]
> I know we discussed the output of failing commands last year, but I don't know the current situation. In general, I think that if a Briefcase command fails because of a subprocess, its output should always be shown to the user unless we can be certain that Briefcase has a more useful message to display instead.

Generally speaking, I agree - however, I think I'd qualify this by saying that the bar for "more useful" should be very low, and skewed in the favor of novice developers. Many of the developer tools we're wrapping leave a lot to be desired in terms of readability; if we're in a position to improve the end-user experience, we should do so.

**Comment by mhsmith** [comment]
Agreed, but in this case, "Unable to install APK" is certainly not a more useful message. Briefcase doesn't know why it failed, but the subprocess output will explain it.

**Comment by freakboy3742** [comment]
Yes, the subprocess output hints at the cause; but it also says that the problem was "ignored" - which... it clearly isn't (at least, not in a way that makes any sense to a Briefcase user). 

I'd suggest that we should inspect the output for the `INSTALL_FAILED_*` marker, and convert that into a more helpful error message if it is detected. If we're able to get a comprehensive list of those failure markers, that would be great; but if not, starting with the 2 we know about an expanding the list as they're reported would also be a significant improvement.

**Comment by moondial-pal** [comment]
To scope this: I'm planning a PR that handles messaging for the two known `INSTALL_FAILED_*` markers (`UPDATE_INCOMPATIBLE` and `VERSION_DOWNGRADE`), matching the existing `INSTALL_FAILED_OLDER_SDK` pattern already in `ADB.run()`.
I suggest leaving the uninstall-then-reinstall flow with a confirmation prompt (raised by @mhsmith) as a separate follow-up, since it's a bigger UX change with its own open questions (confirmation wording, the data-wipe warning). Happy to open a new issue for that once this lands, unless it is preferred to see it bundled in the same PR.

**Comment by freakboy3742** [comment]
@moondial-pal I think that sounds like a good approach - handling those errors *at all* is definitely preferable to nothing. A confirmation workflow will be a lot more complication for what is somewhat of an edge case; as long as it's clear to the end user what action is required in response to getting the error message, that sounds perfectly acceptable.
